services/spar/test/Test/Spar/Sem/ScimExternalIdStoreSpec.hs

```diff
@@ -0,0 +1,165 @@
+{-# LANGUAGE QuantifiedConstraints #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+{-# OPTIONS_GHC -fplugin=Polysemy.Plugin #-}
+
+module Test.Spar.Sem.ScimExternalIdStoreSpec where
+
+import Arbitrary ()
+import Imports
+import Polysemy
+import Polysemy.Check
+import qualified Spar.Sem.ScimExternalIdStore as E
+import Spar.Sem.ScimExternalIdStore.Mem
+import Test.Hspec
+import Test.Hspec.QuickCheck
+import Test.QuickCheck
+
+deriveGenericK ''E.ScimExternalIdStore
+
+propsForInterpreter ::
+  PropConstraints r f =>
+  String ->
+  (forall a. Sem r a -> IO (f a)) ->
+  Spec
+propsForInterpreter interpreter lower = do
+  describe interpreter $ do
+    prop "delete/delete" $ prop_deleteDelete lower
+    prop "delete/lookup" $ prop_deleteLookup lower
+    prop "delete/insert" $ prop_deleteInsert lower
+    prop "lookup/insert" $ prop_lookupInsert lower
+    prop "insert/delete" $ prop_insertInsert lower
+    prop "insert/lookup" $ prop_insertLookup lower
+    prop "insert/insert" $ prop_insertInsert lower
+
+spec :: Spec
+spec = modifyMaxSuccess (const 1000) $ do
+  propsForInterpreter "scimExternalIdStoreToMem" $ pure . run . scimExternalIdStoreToMem
+
+-- | All the constraints we need to generalize properties in this module.
+-- A regular type synonym doesn't work due to dreaded impredicative
+-- polymorphism.
+class
+  (Member E.ScimExternalIdStore r, forall z. Show z => Show (f z), forall z. Eq z => Eq (f z)) =>
+  PropConstraints r f
+
+instance
+  (Member E.ScimExternalIdStore r, forall z. Show z => Show (f z), forall z. Eq z => Eq (f z)) =>
+  PropConstraints r f
+
+prop_insertLookup ::
+  PropConstraints r f =>
+  (forall a. Sem r a -> IO (f a)) ->
+  Property
+prop_insertLookup =
+  prepropLaw @'[E.ScimExternalIdStore] $ do
+    tid <- arbitrary
+    email <- arbitrary
+    uid <- arbitrary
+    pure
+      ( do
+          E.insert tid email uid
+          E.lookup tid email,
+        do
+          E.insert tid email uid
+          pure (Just uid)
+      )
+
+prop_lookupInsert ::
+  PropConstraints r f =>
+  (forall a. Sem r a -> IO (f a)) ->
+  Property
+prop_lookupInsert =
+  prepropLaw @'[E.ScimExternalIdStore] $ do
+    tid <- arbitrary
+    email <- arbitrary
+    pure
+      ( do
+          E.lookup tid email >>= maybe (pure ()) (E.insert tid email),
+        do
+          pure ()
+      )
```

Comment:
The value being equal is tautological, so you're after the fact here that nothing is written to the state if lookup fails?  What's the difference between your test and this one?

```suggestion
      ( E.lookup tid email,
        pure Nothing
      )
```

This question applies to (some of) the other modules as well.

Reply:
This test is that storing what you got is identity. It's the equivalent of this lens law `set l (view l s) s  ≡ s`. In essence it's saying there is no "junk" inside the `insert` method --- eg it's not allowed to change any other keys' values.

Reply:
It's not tautological --- we're also checking equality of the _underlying stores_ after the fact.

Reply:
Ok, I did figure out your second point myself, just needed help with the first one.  Ok, now I want those all to stay, cool!